### PR TITLE
feat(multiphase): Add Phase data to Envoy model

### PR DESCRIPTION
**Pull request by catsmanac**
Extend Envoy model with phase data for Production and Consumption if CT meters are installed and configured.

- Add phase data to production and consumption models and Envoy data class.
- Updated endpoints tests for new phase data
- Updated snapshot with phase data
- Add _common_properties dict attribute and  _add_common_property method to EnvoyUpdater class to share non-binary information between probes or between updaters.
- Add _common_properties property to Envoy class and obtain phase_count property from _common_properties. Pass self._common_properties to probes for registering in EnvoyUpdater class.
- Store meters phaseCount and number of found CT meters info in common_properties

### Production and Consumption Phase data

Only available for metered Envoy with installed and configured CT meter in `three` phase mode and more then 1 phase active. It is reported for solar production and house consumption if these are enabled. EnvoyData now has `system_production_phases: dict[str,EnvoySystemProduction]` and `system_consumption_phases: dict[str,EnvoySystemConsumption]`. Phases are named `L1`, `L2`, and `L3`.
```python
data: EnvoyData = await envoy.update()

if envoy.phase_count > 1:
    for phase in data.system_production_phases:
        print(f'{phase} Watts: {data.system_production_phases[phase].watts_now}')
        print(f'{phase} TodaysEnergy: {data.system_production_phases[phase].watt_hours_today}')
        print(f'{phase} LifetimeEnergy {data.system_production_phases[phase].watt_hours_lifetime}')
        print(f'{phase} Last7DaysEnergy {data.system_production_phases[phase].watt_hours_last_7_days}')
```

### common properties
```python
envoy = Envoy(host)
await envoy.setup()
await envoy.authenticate(username=username, password=password, token=token)

await envoy.probe()
#display firmware version obtained in setup
print(f'Envoy {envoy.host} with {envoy.firmware} {envoy.serial_number}')

#show common_properties reported by updaters probe
print(f'{envoy.common_properties.get("ctMeters")} Meters in {envoy.common_properties.get("phaseMode")} mode')
```

updated phase_count method of envoy now using phase count reported in common_properties by meters probe.

```python
    @property
    def phase_count(self) -> int:
        """Return the number of configured phases."""
        return (
            1
            if "phaseCount" not in self._common_properties
            else self._common_properties["phaseCount"]
        )
```

### CODE REVIEW

File: src/pyenphase/envoy.py [hunk 1]
```diff
@@ -259,18 +260,20 @@ def supported_features(self) -> SupportedFeatures:
...
-            phase_count = 3
-        else:
-            # return 1 phase, indicating no multiphase data is present
-            phase_count = 1
-        return phase_count
+        return (
+            1
+            if "phaseCount" not in self._common_properties
+            else self._common_properties["phaseCount"]
+        )
```
Comment (bdraco) [inline src/pyenphase/envoy.py:276]:
```suggestion
        return self._common_properties.get("phaseCount", 1)
```

File: src/pyenphase/models/system_production.py [hunk 2]
```diff
@@ -58,3 +58,24 @@ def from_production(cls, data: dict[str, Any]) -> EnvoySystemProduction:
...
+    def from_phase(
+        cls, data: dict[str, Any], phase: int
+    ) -> EnvoySystemProduction | None:
+        """Initialize from the production API phase data."""
+        all_production = data["production"]
+
+        eim = find_dict_by_key(all_production, "eim")
+        phases = eim.get("lines")
+        if not phases:
+            return None
```
Comment (bdraco) [inline src/pyenphase/models/system_production.py:72]: Missing coverage

File: src/pyenphase/updaters/base.py [hunk 3]
```diff
@@ -35,6 +37,10 @@ async def _json_probe_request(self, end_point: str) -> Any:
         response = await self._probe_request(end_point)
         return json_loads(end_point, response.content)
 
+    async def _add_common_property(self, property: str, property_value: Any) -> None:
```
Comment (bdraco) [inline src/pyenphase/updaters/base.py:40]: This doesn't need to be a co-routinue, nothing is being awaited here

File: src/pyenphase/updaters/meters.py [hunk 4]
```diff
@@ -31,19 +44,42 @@ async def probe(
...
+        await self._add_common_property("ctMeters", self.ct_Meters_count)
+        await self._add_common_property("phaseMode", self.phase_mode)
+        await self._add_common_property("consumptionMeter", self.consumption_meter_type)
 
-        if phase_count > 2:
+        # report DUAL or THREE PHASE feature for use by next updaters probe
+        if self.phase_count > 2:
             self._supported_features |= SupportedFeatures.THREEPHASE
-        elif phase_count > 1:
+        elif self.phase_mode == "split":
```
Comment (bdraco) [inline src/pyenphase/updaters/meters.py:82]: phase_mode could be an enum

File: src/pyenphase/envoy.py (hunk 1 above)
Comment (catsmanac) [reply src/pyenphase/envoy.py:276]: will do, tnx

File: src/pyenphase/updaters/base.py (hunk 3 above)
Comment (catsmanac) [reply src/pyenphase/updaters/base.py:40]: Tnx, will change

**Comment by bdraco** [comment]
I ran out of time to finish a review. I've been busy with aiohttp PRs so won't have time to come back to this for a while

Hopefully one of the other pyenphase members can finish up the review

**Comment by catsmanac** [comment]
No problem, Thanks for your time, working on the comments now. When done I'll commit the changes.

**Comment by catsmanac** [comment]
CI now worked. Resolved conflicts with main.

File: src/pyenphase/models/meters.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class EnvoyMeteringStatus(StrEnum):
+    NORMAL = "normal"
+    NOTMETERING = "not-metering"
+    CHECKWIRING = "check-wiring"
+
+
+class EnvoyMeterStatusFlags(StrEnum):
+    PODUCTIONIMBALANCE = "production-imbalance"
```
Comment (bdraco) [inline src/pyenphase/models/meters.py:31]:
```suggestion
    PRODUCTION_IMBALANCE = "production-imbalance"
```

Please separate words with `_`

File: src/pyenphase/models/meters.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+class EnvoyMeterState(StrEnum):
+    ENABLED = "enabled"
+    DISABLED = "disabled"
+
+
+class EnvoyMeteringStatus(StrEnum):
+    NORMAL = "normal"
+    NOTMETERING = "not-metering"
+    CHECKWIRING = "check-wiring"
```
Comment (bdraco) [inline src/pyenphase/models/meters.py:27]:
```suggestion
    NOT_METERING = "not-metering"
    CHECK_WIRING = "check-wiring"
```

File: src/pyenphase/models/meters.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+class EnvoyPhaseMode(StrEnum):
+    SPLIT = "split"
+    THREE = "three"
+    SINGLE = "single"
+
+
+class EnvoyMeterType(StrEnum):
+    PRODUCTION = "production"
+    NETCONSUMPTION = "net-consumption"
+    TOTALCONSUMPTION = "total-consumption"
```
Comment (bdraco) [inline src/pyenphase/models/meters.py:16]:
```suggestion
    PRODUCTION = "production"
    NET_CONSUMPTION = "net-consumption"
    TOTAL_CONSUMPTION = "total-consumption"
```

File: src/pyenphase/models/meters.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    NORMAL = "normal"
+    NOTMETERING = "not-metering"
+    CHECKWIRING = "check-wiring"
+
+
+class EnvoyMeterStatusFlags(StrEnum):
+    PODUCTIONIMBALANCE = "production-imbalance"
+    NEGATIVEPRODUCTION = "negative-production"
+    POWERONUNUSEDPHASE = "power-on-unused-phase"
+    NEGATIVECONSUMPTION = "negative-total-consumption"
```
Comment (bdraco) [inline src/pyenphase/models/meters.py:34]:
```suggestion
    NEGATIVE_PRODUCTION = "negative-production"
    POWER_ON_UNUSED_PHASE = "power-on-unused-phase"
    NEGATIVE_TOTAL_CONSUMPTION = "negative-total-consumption"
```

File: src/pyenphase/models/meters.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class EnvoyMeterKeys(StrEnum):
+    EID = "eid"
+    STATE = "state"
+    TYPE = "measurementType"
+    PHASEMODE = "phaseMode"
+    PHASECOUNT = "phaseCount"
+    METERINGSTATUS = "meteringStatus"
+    FLAGS = "statusFlags"
```
Comment (bdraco) [inline src/pyenphase/models/meters.py:44]:
```suggestion
    EID = "eid"
    STATE = "state"
    TYPE = "measurementType"
    PHASE_MODE = "phaseMode"
    PHASE_COUNT = "phaseCount"
    METERING_STATUS = "meteringStatus"
    STATUS_FLAGS = "statusFlags"
```

File: src/pyenphase/updaters/meters.py [hunk 10]
```diff
@@ -14,11 +20,34 @@ class EnvoyMetersUpdater(EnvoyUpdater):
 
     end_point = ENDPOINT_URL_METERS
     data_end_point = ENDPOINT_URL_METERS_READINGS
+    production_meter_eid = None
+    consumption_meter_eid = None
+    production_meter_type: EnvoyMeterType | None = None
+    consumption_meter_type: EnvoyMeterType | None = None
+    phase_mode: EnvoyPhaseMode | None = None
+    phase_count = 1
+    ct_Meters_count = 0
```
Comment (bdraco) [inline src/pyenphase/updaters/meters.py:29]: Please use snake case here
```suggestion
    ct_meters_count = 0
```

File: src/pyenphase/updaters/meters.py [hunk 11]
```diff
@@ -14,11 +20,34 @@ class EnvoyMetersUpdater(EnvoyUpdater):
...
+
+    def _set_my_common_properties(self) -> None:
```
Comment (bdraco) [inline src/pyenphase/updaters/meters.py:31]:
```suggestion
    def _set_common_properties(self) -> None:
```

my is already implied since its set on the object

File: src/pyenphase/models/meters.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
+"""Model for the Envoy's CT Meters."""
+from __future__ import annotations
```
Comment (bdraco) [inline src/pyenphase/models/meters.py:2]: I think the StrEnum is a good change, but I'd do it in its own PR as this PR is already quite large

File: src/pyenphase/const.py [hunk 13]
```diff
@@ -57,3 +57,9 @@ class SupportedFeatures(enum.IntFlag):
     TARIFF = 128
     DUALPHASE = 256
     THREEPHASE = 512
+
+
+class PhaseNames(enum.StrEnum):
+    PHASE1 = "L1"
+    PHASE2 = "L2"
+    PHASE3 = "L3"
```
Comment (bdraco) [inline src/pyenphase/const.py:65]:
```suggestion
    PHASE_1 = "L1"
    PHASE_2 = "L2"
    PHASE_3 = "L3"
```

File: src/pyenphase/models/meters.py (hunk 12 above)
Comment (catsmanac) [reply src/pyenphase/models/meters.py:2]: Understand, but some of these are in the new test cases config in test_endpoints already. I can remove the inused ones and re-add them when needed in next feature.

Comment (bdraco) [reply src/pyenphase/models/meters.py:2]: I'd open a new PR to add the strenums and than once we merge that one, we can merge the changes back into this one so they shake out here and make it smaller to review

Comment (catsmanac) [reply src/pyenphase/models/meters.py:2]: Ah, get it. I'll try to create some smaller steps first.

**Comment by catsmanac** [comment]
Canceling this one as it was too big. Other PR, once #121 is merged will implement feature step by step
